src/main/scala/com/redislabs/provider/redis/streaming/RedisStreamReceiver.scala

```diff
@@ -0,0 +1,199 @@
+package com.redislabs.provider.redis.streaming
+
+import java.util.AbstractMap.SimpleEntry
+
+import com.redislabs.provider.redis.util.Logging
+import com.redislabs.provider.redis.util.PipelineUtils.foreachWithPipeline
+import com.redislabs.provider.redis.{ReadWriteConfig, RedisConfig}
+import org.apache.curator.utils.ThreadUtils
+import org.apache.spark.storage.StorageLevel
+import org.apache.spark.streaming.receiver.Receiver
+import org.spark_project.guava.util.concurrent.RateLimiter
+import redis.clients.jedis.{EntryID, Jedis, StreamEntry}
+
+import scala.collection.JavaConversions._
+
+/**
+  * Receives messages from Redis Stream
+  */
+class RedisStreamReceiver(consumersConfig: Seq[ConsumerConfig],
+                          redisConfig: RedisConfig,
+                          readWriteConfig: ReadWriteConfig,
+                          storageLevel: StorageLevel)
+  extends Receiver[StreamItem](storageLevel) with Logging {
+
+  override def onStart(): Unit = {
+    logInfo("Starting Redis Stream Receiver")
+    val executorPool = ThreadUtils.newFixedThreadPool(consumersConfig.size, "RedisStreamMessageHandler")
+    try {
+      // start consumers in separate threads
+      for (c <- consumersConfig) {
+        executorPool.submit(new MessageHandler(c, redisConfig, readWriteConfig))
+      }
+    } finally {
+      // terminate threads after the work is done
+      executorPool.shutdown()
+    }
+  }
+
+  override def onStop(): Unit = {
+  }
+
+  private class MessageHandler(conf: ConsumerConfig,
+                               redisConfig: RedisConfig,
+                               implicit val readWriteConfig: ReadWriteConfig) extends Runnable {
+
+    val jedis: Jedis = redisConfig.connectionForKey(conf.streamKey)
+    val rateLimiterOpt: Option[RateLimiter] = conf.rateLimitPerConsumer.map(r => RateLimiter.create(r))
+
+    override def run(): Unit = {
+      logInfo(s"Starting MessageHandler $conf")
+      try {
+        createConsumerGroupIfNotExist()
+        receiveUnacknowledged()
+        receiveNewMessages()
+      } catch {
+        case e: Exception =>
+          restart("Error handling message. Restarting.", e)
+      }
+    }
+
+    def createConsumerGroupIfNotExist(): Unit = {
+      try {
+        val entryId = conf.offset match {
+          case Earliest => new EntryID(0, 0)
+          case Latest => EntryID.LAST_ENTRY
+          case IdOffset(v1, v2) => new EntryID(v1, v2)
+        }
+        jedis.xgroupCreate(conf.streamKey, conf.groupName, entryId, true)
+      } catch {
+        case e: Exception =>
+          if (!e.getMessage.contains("already exists")) throw e
```

Comment:
Did you try to run it with multiple workers? how many times is that code going to run in a distributed environment?

Reply:
It will run as many time as receivers/dstreams created. That's why I catch an exception and ignore an error if it's already exists